code/modules/tram/tram.dm

```diff
@@ -0,0 +1,264 @@
+
+/obj/tram/tram_controller/Destroy()
+	for(var/obj/tram/floor/F in tram_floors)
+		remove_floor(F)
+	for(var/obj/tram/wall/TW in tram_walls)
+		remove_wall(TW)
+	for(var/obj/tram/controlpad/CP in controllers)
+		remove_controller(CP)
+	killLoop()
+	return ..()
+
+/obj/tram/tram_controller/emp_act(severity)
+	if(automode)	automode = 0
+	..()
+
+/obj/tram/tram_controller/proc/startLoop()
+	if(activated)	return
+	activated = 1
+	spawn(0)
+		while(activated)
+			process()
+			looptick++
+			sleep(1)
+
+/obj/tram/tram_controller/proc/killLoop()
+	activated = 0
+	looptick = 0
+
+/obj/tram/tram_controller/process()
+	update_tram() //Update combine to account for new mobs and/or objects
+	if(automode)
+		tram_rail_follow()
+		if(fast_mode)
+			tram_rail_follow()
+
+/obj/tram/tram_controller/proc/update_tram()
+	tram.Cut()
+	init_tram()
+
+/obj/tram/tram_controller/proc/tram_rail_follow()
+	var/obj/tram/rail/stored_rail = null
+	if(delay_timer >= world.time)	return
+	for(var/obj/tram/rail/RT in get_turf(src))
+		if(RT.stop_duration && !delay_timer)
+			delay_timer = world.time + RT.stop_duration
+			return
+		if(RT.godir)
+			handle_move(RT.godir)
+			last_played_rail = RT
+			return
+		stored_rail = RT
+	for(var/cdir in cardinal)
+		for(var/obj/tram/rail/R in get_step(src,cdir))
+			if(!istype(R))	continue
+			if(R != last_played_rail)
+				handle_move(get_dir(src,R))
+				last_played_rail = stored_rail
+				return
+
+//INITIALIZATION PROCS
+
+/obj/tram/tram_controller/proc/init_floors()
+	var/turf/T = get_turf(src)
+	if(!T)	return
+	var/obj/tram/floor/TTF = locate(/obj/tram/floor) in T
+	if(istype(TTF))	add_floor(TTF) //Find and link floor on controller turf
+	for(var/cdir in cardinal)
+		var/turf/T2 = get_step(T,cdir)
+		var/obj/tram/floor/TF = locate(/obj/tram/floor) in T2
+		if(istype(TF))
+			if(TF in tram_floors)	continue
+			add_floor(TF) //Find and link all cardinally surrounding floors
+			spawn(1)
+				TF.spread_floors() //Do the same thing over again from the floor itself
+
+/obj/tram/tram_controller/proc/init_walls()
+	var/turf/T = get_turf(src)
+	if(!T)	return
+	if(!tram_floors.len)	return
+	var/obj/tram/floor/TTW = locate(/obj/tram/wall) in T //Find and link wall on controller turf
+	if(istype(TTW))	add_wall(TTW)
+	for(var/obj/tram/floor/TF in tram_floors)
+		for(var/cdir in cardinal)
+			var/obj/tram/wall/TW = locate(/obj/tram/wall) in get_step(TF,cdir)
+			if(istype(TW))
+				if(TW in tram_walls)	continue
+				add_wall(TW) //Find and link all cardinally surrounding walls
+				spawn(1)
+					TW.spread_walls() //Do the same thing over again from the wall itself
+
+/obj/tram/tram_controller/proc/init_tram()
+	tram = tram_floors + tram_walls //Tram is everything that makes up the tram
+	for(var/obj/tram/OT in tram)
+		var/turf/T = get_turf(OT)
+		for(var/atom/movable/AM in T) //Including anything inside of it
+			if(AM in tram)	continue
+			if(!check_validity(AM))	continue
+			tram += AM
+			if(!(src in tram))
+				tram += src
```

Comment:
It means for every AM you will also have a src, doubling the number of items that are not tram_floors or tram_walls...